### PR TITLE
refactor: tests/utils.go cleanup

### PR DESCRIPTION
**Pull request by fossedihelm**
**What this PR does / why we need it**:
`tests/utils.go` file contains some utility functions that can be considered as duplicated.
In fact many things can be achieved with the already present `libvmi`/etc funcions.
Other times, it is possible to extract functions from `utils` to `libvmi`.
This PR aims to extract and refactor a small part of the tests/utils.go file.
In particular, in this PR we:
- extract `libvmi.WithHostDisk`.
- extend `utils.NewRandomVMI` function to accept libvmi options.
- remove `utils.NewRandomVMIWithPVCAndUserData`, `NewInt32`, `NewBool` and `CreateVmiOnNode` functions.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
NONE
```
/sig code-quality

### CODE REVIEW

File: tests/libvmi/storage.go
Comment (dankenigsberg) [inline tests/libvmi/storage.go:25]: I don't think that this is acceptable. `libvmi` should not depend on our testsuite. Ideally, it would become a general-purpose go binding for defining kubevirt VMI - not just in our tests but also elsewhere.

File: tests/libvmi/vmi.go
Comment (dankenigsberg) [inline tests/libvmi/vmi.go:254]: this change of signature is not self-explanatory. Why is it better with `nodeName`?

I have a feeling that your code change could be less "disruptive" if you add `libnode.GetNodeByName()` to be used in few places.

File: tests/utils.go
Comment (dankenigsberg) [inline tests/utils.go:880]: removing code is nice.

File: tests/utils.go
Comment (dankenigsberg) [inline tests/utils.go:543]: I think I understand why you do this, but there's a downside: you may motivate people to keep using this function instead of just replacing it with the relevant `libvmi` calls.

File: tests/storage/restore.go
Comment (dankenigsberg) [inline tests/storage/restore.go:1232]: as mentioned elsewhere, I'd rather see you drop the usage of `tests` completely, by inlining what `NewRandomVMI` does.

File: tests/utils.go
Comment (dankenigsberg) [inline tests/utils.go:1521]: Deleting dead code is noble. But good commits do one thing. I'd put it in a commit of its own.

File: tests/utils.go
Comment (dankenigsberg) [inline tests/utils.go:1048]: heh, I have a pending PR with practically this change.

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

File: tests/libvmi/vmi.go
Comment (EdDev) [reply tests/libvmi/vmi.go:254]: Per what I see, it was changed to support sceanarios where only the name is available (e.g. from the pod) and not the node object.
Fetching the node object from the API at callers seems an overkill.

I do not see a problem with this change, unless in the future other node fields will be used for the affinity. But the future is not clear in this case, so I think we can keep it as is.

File: tests/utils.go
Comment (EdDev) [reply tests/utils.go:543]: I agree, but I would be fine with just a promise to continue this in a follow up to fix it.

My stronger concern here is that it is misleading.
The options are added to the VMI, but later several options are overridden. This is wrong.

Maybe you can try and remove in a dedicated commit:
```go
        vmi.ObjectMeta.Namespace = testsuite.GetTestNamespace(vmi)
	vmi.Spec.Domain.Resources.Requests = k8sv1.ResourceList{}
	if checks.IsARM64(testsuite.Arch) {
		// Cirros image need 256M to boot on ARM64,
		// this issue is traced in https://github.com/kubevirt/kubevirt/issues/6363
		vmi.Spec.Domain.Resources.Requests[k8sv1.ResourceMemory] = resource.MustParse("256Mi")
	} else {
		vmi.Spec.Domain.Resources.Requests[k8sv1.ResourceMemory] = resource.MustParse("128Mi")
	}
```

File: tests/network/macvtap.go
Comment (EdDev) [inline tests/network/macvtap.go:78]: I do not see the usefulness of these helper functions, they just make a mess.

Can you simply inline the building of the VMI?

(BTW, I do not understand the slice appending done here, it converts the options into a slice which at the end expands them).

File: tests/network/macvtap.go
Comment (EdDev) [inline tests/network/macvtap.go:109]: A direct creation of the VMI should be enough. I do not understand why there is a need to retry the creation for 90sec. And it was not there before.

File: tests/network/sriov.go
Comment (EdDev) [inline tests/network/sriov.go:804]: Same, I do not think we need a retry for the creation. This should be now a simple client "create" call.

File: tests/network/sriov.go
Comment (EdDev) [inline tests/network/sriov.go:674]: This does not look good now: Some options are added in this function, some are passed from outside.
But for the ones passed from outside, it is just for a single option.

While I cannot ask to inline this one, as it is used in many places, you could do this small trick:

```go
vmi := newSRIOVVmi([]string{networkName}, cloudInitNetworkDataWithStaticIPsByMac(networkName, mac.String(), cidr))
addNodeAffinity = libvmi.WithNodeAffinityFor(nodeName)
addNodeAffinity(vmi)
```

It is a bit hacky, but IMO much better than the current suggestion.

File: tests/network/vmi_multus.go
Comment (EdDev) [inline tests/network/vmi_multus.go:187]: Same, no retry is needed.

Consider this relevant to all other occurrences in this commit.

File: tests/libvmi/vmi.go
Comment (EdDev) [inline tests/libvmi/vmi.go:258]: Please relocate to a dedicated file (see suggested rules in the readme), vmi.go should not be populated unless there is a very good reason.

This may fit into `selector` or `nodeselector`, or anything you think fits better to describe this group of options.

**Review by EdDev — CHANGES_REQUESTED** [review]
Thank you, nice cleaning.

File: tests/utils.go
Comment (fossedihelm) [reply tests/utils.go]: I opted to drop this change and switch using `libvmi` calls as @dankenigsberg suggested

File: tests/utils.go
Comment (fossedihelm) [reply tests/utils.go:1521]: Done

File: tests/network/macvtap.go
Comment (fossedihelm) [reply tests/network/macvtap.go]: I used the widely used `RunVMI` function, but it is ok for me to switch toward single creation function call.

**Comment by fossedihelm** [comment]
Thanks @dankenigsberg and @EdDev for your review! I addressed the comments. PTAL!

File: tests/storage/storage.go
Comment (EdDev) [inline tests/storage/storage.go:557]: This is not needed, the namespace specified as part of the creation should be enough (line 560).
Is there some other magic that happens here which I missed?

This should be true to all occurences of this in this PR.

BTW, the addition of this option to libvmi is probably incorrect, it needs to be removed.
I think there was only one case when it was needed to be added to the spec, due to some webhook that depended on it (which is a bug IMO).

File: tests/libvmi/factory.go
Comment (EdDev) [inline tests/libvmi/factory.go:90]: This looks wrong to me.
The helper is specific to Cirros factory, if something else needs it, then:
- Use Cirros factory there.
- Copy this logic there.
- Place the upper value there.
- Add a new factory that serves some other common image/guest.

File: tests/network/macvtap.go
Comment (EdDev) [reply tests/network/macvtap.go]: Widely used does not assure it is correct.
Someone needs a good reason to retry creation failure IMO.

The fact that it was not needed before here, hints it is not a proper replacement.

**Review by EdDev — CHANGES_REQUESTED** [review]
Thanks, see comments inline.

Note that the last commit, is adding `tests.RunVMI` and also changing prev usages of it to a direct create. I guess something got mixed there.

File: tests/libvmi/factory.go
Comment (fossedihelm) [reply tests/libvmi/factory.go:90]: To me, this is not wrong. If I want to get the Cirros memory to create a custom cirros vmi I think this is helpful. Also, if I want to use the cirros memory, this does not mean I want to use the entirely cirros spec.
A new helper function like: `BaseMemory()` will be accepted? Consider that currently, cirrosMemory is also used for alpineMemory, that seems even worst to me.
I am not going to duplicate a logic in 10 places.

**Review by fossedihelm — COMMENTED** [review]
Thanks @EdDev!
I am going to re-arrange the commit contents. Maybe smth goes wrong during the squash. In the meanwhile, can you have a look to my comments answers? Thanks

File: tests/libvmi/factory.go
Comment (EdDev) [reply tests/libvmi/factory.go:90]: >  If I want to get the Cirros memory to create a custom cirros vmi I think this is helpful.

If one wants Cirros, why not to use its factory?

> Also, if I want to use the cirros memory, this does not mean I want to use the entirely cirros spec.

Based on the name,  it does. I guess it needs a name change, which I never had.

> A new helper function like: BaseMemory() will be accepted?

The limitation was with Cirros and Alpine, not with all guests/images.
The condition was not added to the base factory, nor to Fedora.

I do not have a better name, I would claim it is per guest OS and therefore should be duplicated.
The fact Cirros and Alpine share the same logic and values, does not mean they depend on the same thing. Even now it looks bad with Alpine sharing the same, but at least it is internal... The proposal here makes this exposed and therefore making this worse IMO.

File: tests/storage/storage.go
Comment (EdDev) [reply tests/storage/storage.go:557]: This is not something I invented really, I got the same comments in the past from others and I think it is improper to specify the namespace as part of the spec.

If the NS specified by the client command collides with the one in the spec, it will blow up.
I do not see a reason to get into that scenario.
The fact the parameter can be provided from two sources is a problem and I would prefer not to encourage it if possible.

The helper you mentioned is using `testsuite.GetTestNamespace` which has the needed logic to work without the NS specified in the spec. BYW, I do not like that helper as well :)

File: tests/libvmi/factory.go
Comment (fossedihelm) [reply tests/libvmi/factory.go]: For example in those cases where you use a PVC with cirros image, or you want a different container disk structure.
Anyway, I think that in most of the cases that was a leftover coming from a previous RandomVMI implementation.
Let's look at the arm e2e job.
PTAL

File: tests/storage/storage.go
Comment (fossedihelm) [reply tests/storage/storage.go:557]: For this refactor I will leave things as is. We can talk about a new refactor maybe in the future, WDYT?

**Comment by fossedihelm** [comment]
close/reopen to trigger dco

File: tests/libvmi/vmi.go
Comment (dankenigsberg) [reply tests/libvmi/vmi.go]: No, please keep the short name. I just wish I saw the motivation declared in a separate commit.

File: tests/storage/storage.go
Comment (dankenigsberg) [reply tests/storage/storage.go:557]: I think you can keep setting the namespace here for a while, but you should not add 
`libvmi.WithNamespace` so as not to confuse others to use it more.

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

File: tests/migration/migration.go
Comment (prnaraya) [inline tests/migration/migration.go:3130]: Nit: I think you can probably combine this comment with the following comment since this is now only referring to labeling the node. Maybe something like this? 

```go
    By("starting a migratable VMI with 2 dedicated CPU cores on the first node of the list")
    libnode.AddLabelToNode(nodes[0].Name, testLabel1, "true")
    vmi := libvmi.NewAlpine(
```

Comment (EdDev) [reply tests/migration/migration.go:3130]: I do not think anything changed in the test logic, so I see no reason to change the logging here.

I usually do not like these comments at all, but some find them useful when reading the test logs (when it fails) to recognize better at what point the failure occurred. Adding the mentioned details will not really add anything valuable to someone that reads the logs, they will still need to visit the code to understand what failed.

**Review by EdDev — APPROVED** [review]
Thank you very much!

The only port which IMO needs to be treated is `libvmi.WithNamespace`.
But as it was already there, and now this is just explicitly showing that it is used, the situation is much better than before.

Therefore, I think we can treat it in a follow up.

Hope to see this merged soon.

File: tests/vmi_configuration_test.go
Comment (xpivarc) [inline tests/vmi_configuration_test.go:2154]: Seems redundant, do we need the guest to boot?

File: tests/storage/restore.go
Comment (xpivarc) [inline tests/storage/restore.go:1241]: Why `libvmi.WithNamespace(testsuite.GetTestNamespace(nil)),`?

**Comment by xpivarc** [comment]
Note: I think the first commit could well be split into 3. Up to you.

File: tests/vmi_configuration_test.go
Comment (fossedihelm) [reply tests/vmi_configuration_test.go:2154]: Maybe you are right! I was just trying to leave things as it was.
But this could save time and cognitive resources.

File: tests/libvmi/vmi.go
Comment (fossedihelm) [reply tests/libvmi/vmi.go]: I will separate the first commit into 3 as @xpivarc suggested. Thanks

File: tests/libvmi/storage.go
Comment (fossedihelm) [reply tests/libvmi/storage.go:106]: For 2 reasons:
- users that want to have the control of the name used
- allow to add multiple host disks

File: tests/libvmi/storage.go
Comment (fossedihelm) [reply tests/libvmi/storage.go:36]: Same as `WithHostDisk`.

**Comment by fossedihelm** [comment]
@xpivarc I addressed (most of) the comments. PTAL
@EdDev FYI

**Review by EdDev — APPROVED** [review]

File: tests/storage/storage.go
Comment (dankenigsberg) [reply tests/storage/storage.go:557]: Oh, I was not aware that `libvmi.WithNamespace` was introduced back in https://github.com/kubevirt/kubevirt/pull/8649

File: tests/libvmi/vmi.go
Comment (dankenigsberg) [reply tests/libvmi/vmi.go]: @xpivarc I don't quite follow you most recent comment. I'd love seeing a commit telling me about the signature change and what it's good for.

**Comment by fossedihelm** [comment]
[Changes](https://github.com/kubevirt/kubevirt/compare/09917b98ebba61aded4f043def106ccd41e5496f..8ef2bab653687e390abb871a6f75cf7006d076d8):
- rebase

**Review by EdDev — APPROVED** [review]

**Comment by xpivarc** [comment]
/approve
/hold
I think it would be safer to merge this after https://github.com/kubevirt/kubevirt/pull/11005 but feel free to unhold.

**Comment by dankenigsberg** [comment]
> I think it would be safer to merge this after https://github.com/kubevirt/kubevirt/pull/11005 but feel free to unhold.

I hope that one day `libvmi` would be used in production code, too. IMO, helper libraries should not `panic()`. Regardless, I hope improvements to `libvmi` can continue in parallel to that discussion. I hope someone unholds https://github.com/kubevirt/kubevirt/pull/10973.

/lgtm

**Comment by fossedihelm** [comment]
@xpivarc I will unhold this since there are other improvements that deserve to get in.
If totally disagree, feel free to unhold again :) 
/unhold

**Pull request merged**
